Data/ByteString/Internal/Pure.hs

```diff
@@ -0,0 +1,363 @@
+encodeSignedDec !x !buf
+  | x >= 0    = encodeUnsignedDec x buf
+  | otherwise = do
+    -- we cannot negate directly as  0 - (minBound :: Int) = minBound
+    -- So we write the sign and the first digit.
+    pokeByteOff buf 0 '-'
+    let !(q,r) = quotRem x (-10)
+    putDigit buf 1 (fromIntegral (abs r))
+    case q of
+      0 -> pure (plusPtr buf 2)
+      _ -> encodeUnsignedDec' q (plusPtr buf 1) (plusPtr buf 2)
+
+
+-- | Decode positive number as decimal
+encodeUnsignedDec :: (Eq a, Num a, Integral a) => a -> Ptr Word8 -> IO (Ptr Word8)
+{-# INLINABLE encodeUnsignedDec #-} -- for specialization
+encodeUnsignedDec !v !next_ptr = encodeUnsignedDec' v next_ptr next_ptr
+
+-- | Decode positive number as little-endian decimal, then reverse it.
+--
+-- Take two pointers (orig_ptr, next_ptr) to support already encoded digits
+-- (e.g. used by encodeSignedDec to avoid overflows)
+--
+encodeUnsignedDec' :: (Eq a, Num a, Integral a) => a -> Ptr Word8 -> Ptr Word8 -> IO (Ptr Word8)
+{-# INLINABLE encodeUnsignedDec' #-} -- for specialization
+encodeUnsignedDec' !v !orig_ptr !next_ptr = do
+  let !(q,r) = divMod v 10
+  putDigit next_ptr 0 (fromIntegral r)
+  case q of
+    0 -> do
+      -- reverse written digits
+      reverseBytesInplace orig_ptr next_ptr
+      -- return pointer after our digits
+      pure (plusPtr next_ptr 1)
+    _ -> encodeUnsignedDec' q orig_ptr (plusPtr next_ptr 1)
+
+encodeUnsignedDecPadded :: (Eq a, Num a, Integral a) => Int -> a -> Ptr Word8 -> IO ()
+{-# INLINABLE encodeUnsignedDecPadded #-} -- for specialization
+encodeUnsignedDecPadded !max_width !v !buf = assert (max_width > 0) $ do
+  let !(q,r) = divMod v 10
+  putDigit buf (max_width - 1) (fromIntegral r)
+  case q of
+    0 -> do
+        -- pad beginning
+        let pad 0 = pure ()
+            pad n = putDigit buf (n - 1) 0 >> pad (n - 1)
+        pad (max_width - 1)
+    _ -> encodeUnsignedDecPadded (max_width - 1) q buf
+
+
+
+-- | Decode positive number as hexadecimal
+encodeUnsignedHex :: (Eq a, Num a, Integral a, Bits a) => a -> Ptr Word8 -> IO (Ptr Word8)
+{-# INLINABLE encodeUnsignedHex #-} -- for specialization
+encodeUnsignedHex !v !next_ptr = encodeUnsignedHex' v next_ptr next_ptr
+
+-- | Decode positive number as little-endian hexdecimal, then reverse it.
+--
+-- Take two pointers (orig_ptr, next_ptr) to support already encoded digits
+encodeUnsignedHex' :: (Eq a, Num a, Integral a, Bits a) => a -> Ptr Word8 -> Ptr Word8 -> IO (Ptr Word8)
+{-# INLINABLE encodeUnsignedHex' #-} -- for specialization
+encodeUnsignedHex' !v !orig_ptr !next_ptr = do
+  -- (q,r) = divMod v 16, but faster
+  let !q = v `shiftR` 4
+  let !r = v .&. 0x0F
+  putDigit next_ptr 0 (fromIntegral r)
+  case q of
+    0 -> do
+      -- reverse written digits
+      reverseBytesInplace orig_ptr next_ptr
+      -- return pointer after our digits
+      pure (plusPtr next_ptr 1)
+    _ -> encodeUnsignedHex' q orig_ptr (plusPtr next_ptr 1)
+
+
+lower_hex_table :: Ptr ()
```

Comment:
`Ptr ()` is a strange type to give this. I'd prefer `Ptr Word8` or perhaps `Ptr Word16` if you really want to emphasize that these are digit pairs. (Either way it should match `digit_pairs_table below`.)